src/tls/README.md

```diff
@@ -0,0 +1,65 @@
+cockpit TLS proxy
+=================
+
+`cockpit-tls` is a TLS terminating HTTP proxy for cockpit-ws. TLS termination
+should not be done directly in `cockpit-ws`, as it is hard to audit and trust:
+This code is subject to external attacks through HTTP, does a *lot* of
+interpretation of data streams, uses a lot of external dependencies, and
+multiplexes sessions of different users in one process. Thus any vulnerability
+can easily lead to complete privilege escalation across all running and future
+sessions.
+
+Right now this does not have any extra features over using cockpit-ws directly
+(other than better isolation). In the future, it will be able to do certificate
+based client authentication (smart cards, browser-imported certificates, and
+similar), which greatly aggravates the above security trust issue.
+
+Design goals
+------------
+
+ * Run TLS termination (cockpit-tls) and the actual data interpretation and
+   sessions (cockpit-ws) in separate security contexts, so that vulnerabilities
+   in the HTTP session cannot affect the trust put into the TLS authentication.
+ * Run separate cockpit-ws processes for each TLS client certificate, plus one
+   instance for "https without certificate", and another one for "unencrypted
+   http", so that one tampered session cannot affect the others.
+ * cockpit-tls *does not* interpret the HTTP data stream, to avoid risking any
+   vulnerability. It is only allowed to inspect the first byte of a new
+   connection to decide between TLS or plain HTTP, and do the TLS negotiation
+   and encryption/decryption. In other words, it treats the payload on the TLS
+   connection as a black box.
+
```

Comment:
Another important goal is to not use any "big" libraries like glib, right?

Reply:
Right, I added that (locally, won't push for just that).